app/code/Magento/Sales/Controller/AbstractController/PrintInvoice.php

```diff
@@ -17,64 +25,95 @@ abstract class PrintInvoice extends \Magento\Framework\App\Action\Action
+     * @var Registry
      */
-    protected $_coreRegistry;
+    protected $coreRegistry;
 
     /**
      * @var PageFactory
      */
     protected $resultPageFactory;
 
+    /**
+     * @var InvoiceRepositoryInterface
+     */
+    protected $invoiceRepository;
+
+    /**
+     * @var Session
+     */
+    protected $session;
+
+    /**
+     * @var Order
+     */
+    protected $order;
+
     /**
      * @param Context $context
      * @param OrderViewAuthorizationInterface $orderAuthorization
-     * @param \Magento\Framework\Registry $registry
+     * @param Registry $registry
      * @param PageFactory $resultPageFactory
      */
     public function __construct(
-        Context $context,
+        Context                         $context,
         OrderViewAuthorizationInterface $orderAuthorization,
-        \Magento\Framework\Registry $registry,
-        PageFactory $resultPageFactory
+        Registry                        $registry,
+        PageFactory                     $resultPageFactory,
+        InvoiceRepositoryInterface      $invoiceRepository,
+        Session                         $session,
+        Order                           $order
     ) {
         $this->orderAuthorization = $orderAuthorization;
-        $this->_coreRegistry = $registry;
+        $this->coreRegistry = $registry;
         $this->resultPageFactory = $resultPageFactory;
+        $this->invoiceRepository = $invoiceRepository;
+        $this->session = $session;
+        $this->order = $order;
         parent::__construct($context);
     }
 
     /**
      * Print Invoice Action
      *
-     * @return \Magento\Framework\Controller\Result\Redirect|\Magento\Framework\View\Result\Page
+     * @return Redirect|Page
      */
     public function execute()
     {
         $invoiceId = (int)$this->getRequest()->getParam('invoice_id');
         if ($invoiceId) {
-            $invoice = $this->_objectManager->create(
-                \Magento\Sales\Api\InvoiceRepositoryInterface::class
-            )->get($invoiceId);
+            try {
+                $invoice = $this->invoiceRepository->get($invoiceId);
+            } catch (NoSuchEntityException $e) {
+                $this->messageManager->addErrorMessage(__($e->getMessage()));
+                /** @var Redirect $resultRedirect */
+                $resultRedirect = $this->resultRedirectFactory->create();
+                if ($this->session->isLoggedIn()) {
+                    $resultRedirect->setPath('*/*/history');
+                } else {
+                    $resultRedirect->setPath('sales/guest/form');
+                }
+                return $resultRedirect;
+            }
             $order = $invoice->getOrder();
         } else {
             $orderId = (int)$this->getRequest()->getParam('order_id');
-            $order = $this->_objectManager->create(\Magento\Sales\Model\Order::class)->load($orderId);
+            $order = $this->order->load($orderId);
```

Comment:
Could you change `\Magento\Sales\Model\Order` to `\Magento\Sales\Api\OrderRepositoryInterface`